### PR TITLE
Add PHP 8.1 compatibility

### PR DESCRIPTION
**Pull request by karyna-t**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | yes
| New Feature   | no
| RFC           | yes/no
| QA            | yes

### Description

Add PHP 8.1 support to the composer.json and update require section with the new dependency versions which support 8.1.

Update several methods with strict typing to make sure that they will be used in the same manner as new RFC requires:
https://wiki.php.net/rfc/deprecate_null_to_scalar_internal_arg

Fixing the test which is failing with php 8.1 installed.
(`Error was PHP Deprecated:  rawurlencode(): Passing null to parameter #1 ($string) of type string is deprecated in laminas-oauth/src/Http/Utility.php on line 194`)

### CODE REVIEW

File: src/Token/AbstractToken.php [hunk 1]
```diff
@@ -122,7 +122,7 @@ public function getTokenSecret()
      * @param  string $value
      * @return \Laminas\OAuth\Token\AbstractToken
      */
-    public function setParam($key, $value)
+    public function setParam($key, string $value)
```
Comment (Ocramius) [inline src/Token/AbstractToken.php]: These changes are BC breaks, since they break child classes

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: src/Token/AbstractToken.php (hunk 1 above)
Comment (karyna-t) [reply src/Token/AbstractToken.php]: @Ocramius  rolled back

**Comment by xmav** [comment]
@Ocramius Is there anything else to fix before merge ?

**Comment by Ocramius** [comment]
@xmav yes: the `composer.lock` got updated with the wrong system dependencies, and needs to be re-generated.

In order to merge, all CI checks must pass here.

File: composer.json [hunk 2]
```diff
@@ -23,9 +23,9 @@
     "require": {
         "php": "^7.3 || ~8.0.0 || ~8.1.0",
         "laminas/laminas-config": "^3.7",
-        "laminas/laminas-crypt": "^3.8",
+        "laminas/laminas-crypt": "~3.6.0",
         "laminas/laminas-http": "^2.15",
-        "laminas/laminas-i18n": "^2.16",
+        "laminas/laminas-i18n": "~2.13.0",
```
Comment (Ocramius) [inline composer.json:28]: I think these should stay as they are.

What we could do is setting `config.platform.php: "7.3.99"`, so it knows it needs to do a `composer update` with PHP 7.3 compatibility.

Dropping PHP 7.3 may be necessary too (bumping to 7.4)

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: src/Http/Utility.php [hunk 3]
```diff
@@ -120,7 +120,7 @@ public function toAuthorizationHeader(array $params, $realm = null, $excludeCust
      */
     public function sign(
         array $params,
-        $signatureMethod,
+        string $signatureMethod,
```
Comment (Ocramius) [inline src/Http/Utility.php]: We can't really make a stricter signature here: it would be a BC break for child classes

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: src/Http/Utility.php (hunk 3 above)
Comment (karyna-t) [reply src/Http/Utility.php]: Hello @Ocramius , I've added this to ensure that there will be no potential deprecation issues because of the non-strict type (since php 8.1 explode() now throws an error).
The same for `urlEncode(string $value)`

Maybe we should an additional checking instead? Something like
`$value !== null ? rawurlencode($value) : '';`

Or you think it's better just leave it as it was before?

Comment (Ocramius) [reply src/Http/Utility.php]: > no potential deprecation issues because of the non-strict type (since php 8.1 explode() now throws an error).

What deprecation warning This signature is entirely defined in this package :thinking: 

If this is about `null` being passed to `$signatureMethod`, we need a `(string)` cast in the body of the method, rather than an added type declaration.

Sadly, adding type declarations is considered a major BC breakage, so its "cost" is very high.

Comment (Ocramius) [reply src/Http/Utility.php:123]: Ah yes, I even voted on that RFC, but it was implemented in a terrible way :|

File: test/OAuthTest.php [hunk 4]
```diff
@@ -154,7 +158,9 @@ public function testOauthClientUsingPostAndGetRequestParametersForSignature()
               ->setTokenSecret('456');
 
         $client = new OAuthClient([
-            'token' => $token
+            'token' => $token,
+            'consumerSecret' => '', // for PHP 8.1 compatibility
```
Comment (Ocramius) [inline test/OAuthTest.php]: These added keys shouldn't be necessary for PHP 8.1 compatibility: can we somehow  change our sources so this doesn't become a BC issue? 🤔

Comment (karyna-t) [reply test/OAuthTest.php]: As we decided to add just a string cast instead of changing signature, these keys are no longer required. I removed them

**Review by Ocramius — APPROVED** [review]
Looking good, thanks @karyna-tsymbal-atwix!

**Comment by Ocramius** [comment]
Released: https://github.com/laminas/laminas-oauth/releases/tag/2.3.0

File: composer.json [hunk 5]
```diff
@@ -15,22 +15,25 @@
...
-        "laminas/laminas-crypt": "^3.0",
-        "laminas/laminas-http": "^2.14",
-        "laminas/laminas-i18n": "^2.11",
-        "laminas/laminas-loader": "^2.7",
-        "laminas/laminas-math": "^3.3",
-        "laminas/laminas-stdlib": "^3.2",
-        "laminas/laminas-uri": "^2.8"
+        "php": "^7.3 || ~8.0.0 || ~8.1.0",
+        "laminas/laminas-config": "^3.7",
+        "laminas/laminas-crypt": "~3.6.0",
```
Comment (Ocramius) [inline composer.json:26]: Ah, I need to change these to `^` - will do a separate patch.

File: composer.json [hunk 6]
```diff
@@ -15,22 +15,25 @@
...
+        "laminas/laminas-http": "^2.15",
+        "laminas/laminas-i18n": "~2.13.0",
```
Comment (Ocramius) [inline composer.json:28]: Ah, I need to change these to `^` - will do a separate patch.
